README.md
## toStringPretty

Prints an object pretty.

## Installing

```yaml
dependencies:
  to_string_pretty:
```

```dart
import 'package:to_string_pretty/to_string_pretty.dart';
```

## Usage

- **Basic**

  ```dart
  var data = ['google', 'apple', 'thanks'];
  print(toStringPretty(data));
  ```

  ```
  // stdout
  [
      google,
      apple,
      thanks,
  ],
  ```

- **toString()** 

  [**Code**](https://github.com/pubmskim/to_string_pretty/blob/main/example/to_string_pretty_example.dart)

  ```dart
  class ChatMessage {
    // member variables, methods...
    
    @override
    String toString() => toStringPretty(this, {
      'body': message
      'sentBy': sentBy,
      'sentAt': sentAt,
      'group': group,
    });
  }
  
  print(ChatMessage());
  ```

  ```
  // stdout
  ChatMessage {
      body: Thanks,
      sentBy: ChatUser {
          email: hello@example.com,
          name: Liam,
      },
      sentAt: ChatDateTime {
          datetime: 2021-02-21 19:07:00.000,
      },
      group: Flutter Community,
  },
  ```






## Feature requests and Bug reports

**[here](https://github.com/pubmskim/to_string_pretty/issues)**

CHANGELOG.md
## 1.0.0

- Release 1.0.0


test/main.dart
import 'package:test/test.dart';

import 'to_string_pretty/test.dart';

void main() {
  group('[toStringPretty()]', testToStringPretty);
}


test/to_string_pretty/test.dart
import 'package:test/test.dart';
import 'package:to_string_pretty/src/impl.dart';

import 'complex1.dart' as complex;
import 'simple.dart' as simple;

void testToStringPretty() {
  test('simple', () {
    final data = simple.Simple(1, '2');
    expect(data.toString(), simple.expected);
  });

  test('complex', () {
    final data = complex.Complex();
    expect(data.toString(), complex.expected);
  });

  test('empty container', () {
    final testCase = [
      [[], '[]'],
      [{}, '{}'],
    ];
    for (final t in testCase) {
      final formatted = toStringPretty(t[0]);
      final expected = t[1];
      expect(formatted, expected);
    }
  });
}


test/to_string_pretty/simple.dart
import 'dart:io';

import 'package:to_string_pretty/src/impl.dart';

class Simple {
  Simple(this.value1, this.value2);
  final int value1;
  final String value2;

  @override
  String toString() => toStringPretty(this, {
        'value1': value1,
        'value2': value2,
      });
}

final expected = File('test/to_string_pretty/simple.txt')
    .readAsStringSync()
    .replaceAll('\r', '');


lib/src/impl.dart
/// Formats the desired object pretty.
///
/// [Example:]
/// ```dart
/// class Inner {
///   Inner();
///   final three = 'clap!';
///
///   @override
///   String toString() => toStringPretty(this, {
///       'three': three,
///     });
/// }
///
/// class Outer {
///   Outer();
///   final one = 1;
///   final two = 2;
///   final three = Inner();
///
///   @override
///   String toString() => toStringPretty(this, {
///       'one': one,
///       'two': two,
///       'three': three,
///     });
/// }
///
/// void main() {
///   print(Outer());
/// }
/// ```
///
/// [Console:]
/// ```bash
///Outer {
///    one: 1
///    two: 2
///    three: Inner {
///        three: clap!
///    },
///},
/// ```
String toStringPretty(Object object, [Map<String, Object> values]) {
  if (object is Iterable) return _toPrettyIterable(object);
  if (object is Map) return _toPrettyMap(object);
  if (object is Set) return _toPrettySet(object);

  final buf = StringBuffer();

  if (object is String) {
    buf.writeln('$object {');
  } else {
    buf.writeln('${object.runtimeType} {');
  }
  values?.forEach((key, value) {
    buf.writeAll(_shiftln(value, key));
  });
  buf.write('},');

  return buf.toString();
}

Iterable<String> _shiftln(Object obj, [String name]) sync* {
  String str;
  if (obj is Iterable) {
    str = '$name: ${_toPrettyIterable(obj)}';
  } else if (obj is Map) {
    str = '$name: ${_toPrettyMap(obj)}';
  } else if (obj is Set) {
    str = '$name: ${_toPrettySet(obj)}';
  } else if (name != null) {
    str = '$name: $obj';
  } else {
    str = obj.toString();
  }

  for (final item in str.split('\n')) {
    if (item.endsWith('{') || item.endsWith('[')) {
      yield '    $item\n';
    } else {
      yield '    ${item.replaceFirst(',', '')},\n';
    }
  }
}

String _toPrettyIterable(Iterable iter) {
  if (iter.isEmpty) return '[]';

  final buf = StringBuffer();
  buf.writeln('[');
  for (final item in iter) {
    if (item is Iterable) {
      buf.writeAll(_shiftln(_toPrettyIterable(item)));
    } else if (item is Map) {
      buf.writeAll(_shiftln(_toPrettyMap(item)));
    } else if (item is Set) {
      buf.writeAll(_shiftln(_toPrettySet(item)));
    } else {
      buf.writeAll(_shiftln(item.toString()));
    }
  }
  buf.write('],');
  return buf.toString();
}

String _toPrettySet(Set iter) {
  if (iter.isEmpty) return '{}';

  final buf = StringBuffer();
  buf.writeln('{');
  for (final item in iter) {
    if (item is Iterable) {
      buf.writeAll(_shiftln(_toPrettyIterable(item)));
    } else if (item is Map) {
      buf.writeAll(_shiftln(_toPrettyMap(item)));
    } else if (item is Set) {
      buf.writeAll(_shiftln(_toPrettySet(item)));
    } else {
      buf.writeAll(_shiftln(item.toString()));
    }
  }
  buf.write('},');
  return buf.toString();
}

String _toPrettyMap(Map map) {
  if (map.isEmpty) return '{}';

  final buf = StringBuffer();
  buf.writeln('{');
  map.forEach((key, value) {
    var str = '$key: ';
    if (value is Iterable) {
      str += _toPrettyIterable(value);
    } else if (value is Map) {
      str += _toPrettyMap(value);
    } else if (value is Set) {
      buf.writeAll(_shiftln(_toPrettySet(value)));
    } else {
      str += value.toString();
    }
    buf.writeAll(_shiftln(str));
  });
  buf.write('},');
  return buf.toString();
}


example/to_string_pretty_example.dart
import 'package:to_string_pretty/to_string_pretty.dart';
import 'package:meta/meta.dart';

class ChatMessage {
  final String message = 'Thanks';
  final ChatUser sentBy = ChatUser();
  final ChatDateTime sentAt = ChatDateTime(DateTime(2021, 02, 21, 19, 07));
  final String group = 'Flutter Community';

  @override
  String toString() => toStringPretty(this, {
        'body': message,
        'sentBy': sentBy,
        'sentAt': sentAt,
        'group': group,
      });
}

class ChatUser {
  final email = 'hello@example.com';
  final name = 'Liam';
  final _isNameVisible = true;

  ChatUser();

  @override
  String toString() => toStringPretty(this, {
        'email': email,
        if (_isNameVisible) 'name': name,
      });
}

class ChatDateTime {
  final DateTime datetime;

  ChatDateTime(this.datetime);

  @override
  String toString() => toStringPretty(this, {
        'datetime': datetime.toLocal(),
      });
}

void main() {
  print(ChatMessage());
}
